### PR TITLE
Add support for marking whole notification stacks as read

### PR DESCRIPTION
**Pull request by notbakaneko**
Allows entire notification stacks to be marked as read in order to support the fix for #4896
The format for this (endpoint is still the same) is 
```ts
{
  object_type: string;
  object_id?: number;
  category?: string;
}
```

~~- [ ] database index needs to be added on `is_read`~~ didn't seem to have any effect on the query ?_?
- [x] unread count is wrong when an entire stack is marked as read

### CODE REVIEW

File: app/Http/Controllers/NotificationsController.php [hunk 1]
```diff
@@ -173,4 +183,48 @@ private function endpointUrl()
 
         return $url;
     }
+
+    private function markReadByIds($ids)
+    {
+        $user = auth()->user();
+        if ($user->userNotifications()->whereIn('notification_id', $ids)->update(['is_read' => true])) {
+            event(new NotificationReadEvent($user->getKey(), $ids));
```
Comment (nanaya) [inline app/Http/Controllers/NotificationsController.php]: might as well pass `['ids'=>$ids]` for `$params` and remove the second argument?

File: app/Http/Controllers/NotificationsController.php [hunk 2]
```diff
@@ -173,4 +183,48 @@ private function endpointUrl()
...
+
+            return response(null, 204);
+        }
+
+        return response(null, 422);
+    }
+
+    private function markReadByNotificationIdentifier($params)
```
Comment (nanaya) [inline app/Http/Controllers/NotificationsController.php]: maybe should look into updating `Jobs\MarkNotificationsRead`...?

File: resources/assets/lib/notifications/notification-identity.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+export interface NotificationIdentityJson {
+  category?: string;
+  id?: number;
+  object_id?: number;
+  object_type: string;
+}
+
+export function resolveStackId(identity: NotificationIdentity) {
```
Comment (nanaya) [inline resources/assets/lib/notifications/notification-identity.ts]: not used?

File: resources/assets/lib/notifications/notification-identity.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  return `${identity.objectType}-${identity.objectId}-${identity.category}`;
+}
+
+export function resolveIdentityType(identity: NotificationIdentity): NotificationIdentityType | undefined {
```
Comment (nanaya) [inline resources/assets/lib/notifications/notification-identity.ts]: might as well make a class?

File: resources/assets/lib/notifications/notification-identity.ts (hunk 3 above)
Comment (notbakaneko) [reply resources/assets/lib/notifications/notification-identity.ts]: is used for notification history page

File: resources/assets/lib/notifications/notification-identity.ts (hunk 4 above)
Comment (notbakaneko) [reply resources/assets/lib/notifications/notification-identity.ts]: needs to work on interfaces

**Pull request merged**